doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,108 @@
+- Feature name: Envelope upsert with order by timestamp (limited scope)
+- Associated: https://github.com/MaterializeInc/materialize/issues/16512
+
+# Summary
+
+- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with a `DISTINCT ON` view to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
+- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` metadata field. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.
+
+# Scope
+
+The scope will be limited to only allow `ORDER BY TIMESTAMP` (ascending order).
+
+- For kafka upsert style topics, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to keep the same behavior it would make sense to limit the `ORDER BY` columns to only the metadata columns.
```

Comment:
What's an upsert-style topic, and how is it distinct from a key-compacted topic?

(We've declared the latter out of scope, which I think is wise. But Kafka itself doesn't ascribe any semantics to null values outside the key-compacted case, and it's not clear to me that it's ever actually a thing people do.)

Reply:
By "upsert style" I meant topics having the semantics so that it could be compacted on some value for a key. 

I would imagine key-compacted to be a kafka specific definition where it can be only compacted on the latest offset value. I guess from that POV, the "upsert style" is more of a looser definition that it can have a different field to compact things by (in this case, timestamp).

Regarding ascribing meaning to null values, it's easier to borrow the definition of tombstones since it already works in a similar fashion for current upsert?

Reply:
Yeah, I don't think I disagree with any of that! But as-is it's a bit of a circular definition.

It seems possible that we could avoid introducing the concept of an upsert-style topic at all:

```suggestion
- In a key-compacted Kafka copic, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to borrow that tombstone semantic for our more flexible upsert source, it would make sense to limit the `ORDER BY` columns to only the metadata columns.
```

Otherwise we should define it! (But it seems like we're inventing something instead of adopting something that we found "in the wild", so I'd be inclined to skip it.)

Reply:
Ah I see what you mean. This makes more sense 👍